### ISSUE TITLE
Support envfile for tasks.json

### ISSUE THREAD

**Issue by felixfbecker** [Microsoft/vscode#36769]
The node and Go debuggers both support specifying an env file for environment variables instead of the `env` hash. It would be great if we could use the same file in tasks.json to have a single source of truth that can also be used by non-vscode scripts

**Comment by dbaeumer** [comment]
Can you provide me with an example on how this should look like. Why isn't this a normal file

**Comment by dbaeumer** [comment]
> Not sure what you mean with "Why isn't it a normal file"?

If it is a simple file what is preventing you from using it in a tasks.json file.

**Comment by felixfbecker** [comment]
If you mean why am I not copying the settings over, it is to have a single source of truth. It's the same justification why node debugger and the Go debugger added this config option. tasks.json is not understood by other tools and not every dev uses VS Code, but an env file can be read in easily by a shell script.

**Comment by dbaeumer** [comment]
I am still puzzled with this. All tasks does is calling other programs. Could the task description not simple pass the right argument (the envfile) to the program it starts. Or do you want tasks to have a setting to read envs vars from a file when you do variable substitution like `${env:xxxx}`?

**Comment by felixfbecker** [comment]
A lot of commands that are candidates to be run through the task system are configured through environment variables - which is why the tasks.json `env` option is great.
I can write a tasks.json that has a bunch of tasks that are interconnected with `dependsOn` and have individual problem matchers and they receive the same env vars.

But not everyone uses VS Code, so you will want to provide a way to run these tasks without VS Code, e.g. from a bash script that runs all these commands and sets the same env vars. You want to avoid fragmenting all these env vars between the tasks.json and the script.

Of course, you could now configure your tasks.json to run just a single bash script instead of all the individual commands. But then you lose the ability to specify a problem matcher per task, run them individually etc.

So it would be great to be able to point tasks.json to an external file that contains env vars as simple `key=value` lines and have the script also read in that external file to have a single source of truth.

**Comment by dbaeumer** [comment]
Now I got it.

**Comment by felixfbecker** [comment]
Another reason for this if your application needs secret environment variables like tokens - you can set them in a .gitignored envfile, but commit the tasks.json

**Comment by myrup** [comment]
I'm missing the envFile as an option for a .Net Core Launch configuration (for .gitignore reasons). Shouldn't it be available for all configurations and tasks that start processes?

**Comment by felixhayashi** [comment]
I agree with @myrup, the `.env` constants should be equally accessible from within `launch.json` – not just  `tasks.json`. I saw there was some similar discussion here in https://github.com/Microsoft/vscode/issues/14523 resulting in a `envFile` property but doesn't seem to have been implemented...

**Comment by petermetz** [comment]
Another use-case why this would be great: You want to store the tasks.json file under version control, but not your environment variables which may contain sensitive data. Without the envFile property there is no easy way to separate the two.

**Comment by cdarwin** [comment]
For the time being I am working around this by adding a prefix to my task command which essentially sources the `.env` file used by `launch.json`.

**Comment by GOST-UA** [comment]
Is this issue still actual?

**Comment by doomleika** [comment]
Would like to have this to, in the mean time. use `$(cat .env | xargs) | <your command>`

**Comment by ekkis** [comment]
I support this request.  having all my environment variables in a file and being able to run tasks with those values available would be very valuable

**Comment by ekkis** [comment]
in fact, I'll go further, having the "env" key in options makes no sense at all.  it means I have to hard-code all the environment variables I need into a VSCode configuration file?? that's a terrible approach

**Comment by adriantorrie** [comment]
Need `.env` file support in tasks.

1. Keeps secrets out of git
2. Define my env vars in one place
3. It is a standard workflow for multiple languages to use `.env` file

Here is another use case:

- I was hoping to recreate this `Makefile` ([link](https://github.com/garethr/multi-stage-build-example/blob/master/Makefile)) in `tasks.json` to prevent having to install make on Windows
- However I need to pass build args from my `.env` file (The docker extension for VSCode isn't flexible enough for this approach)
 - Because I cant do the above: I now have to write separate `docker-compose.<make step>.yml` files in the meantime
  - But: `docker-compose` doesn't support `BuildKit` yet, which means I can't bind mount local volumes into the build scope, which would allow me to export unit test results from the images at *build* time (also don't get the parallel stage builds either)
 - Some of the runtime arguments are secrets (i.e. `docker run -e password=$PASSWORD`), so there is no way they are being hard-coded into `tasks.json`
  - So I don't have any means to a fully-functional multi-stage build with build-time bind mounts

The above workflow example in the `Makefile` will be reused in our CI pipeline.
- Essentially allows our developers to run locally what will be run in the CI pipeline
- This gives us a CI pipeline agnostic approach that prevents vendor lock-in

So the only thing we're missing at the moment is being able to give devs a *clean* workflow on their local computers.

** **/s** ** Or we tell them them to maintain both `.env` and `tasks.json` themselves and keep them out of git, and that we expect them to  do this for the 50+ microservices we have built in the past 8 months, and to keep doing it for the additional ones in the future.

**Comment by ekkis** [comment]
> For the time being I am working around this by adding a prefix to my task command which essentially sources the `.env` file used by `launch.json`.

how do you do it? my .env file doesn't include the "export" keyword and thus `. .env` doesn't work as child processes can't see the variables.  I also don't see any parameters to the `source` command to automatically export

**Comment by warnes** [comment]
A nice secondary feature would be to include an `overwrite:True` argument `envFile` to control whether the settings in the file overwrite existing environment variable values.

**Comment by ayoubserti** [comment]
Hi is there any progress on this request?
I need also to define some `.env` variable locally to be used by `tasks.json`

**Comment by dellagustin-sap** [comment]
This seems to going in a similar direction to #89825

**Comment by alexandrulesi** [comment]
I'd also be interested in seeing this. 

> This seems to going in a similar direction to #89825

The solution suggested in the other ticket is to add inputs to tasks.json, but what we are ultimately looking for is a way to run the debugger and only update all environment variables in `.env`, and not in multiple places.

**Comment by mpelnens** [comment]
This would be extremely useful in cases environment variables contain sensitive data like tokens or passwords. With 'envFile' support the .env file could be added to .gitignore

**Comment by pcherna** [comment]
In light testing, the following seems to be an OK workaround for Linux/Mac. Create a .env file with your local config and secrets, e.g.:

```
FOO_USERNAME=person@example.com
FOO_PASSWORD=SecretDoNotShare
```

Then in your tasks.json, where before you had this bad solution:

```
...
"command": "myfoo person@example.com:SecretDoNotShare ...
...
```

you can now say:

```
...
"command": "export $(xargs < .env) && myfoo $FOO_USERNAME:$FOO_PASSWORD ...
...
```

**Comment by zkldi** [comment]
Any progress on this? Seems like a fairly necessary feature for tasks.

**Comment by joundso** [comment]
Aren't they implemented now?
At least something like this is working:
```json
{
  "version": "2.0.0",
  "tasks": [
    {
      "type": "docker-build",
      "label": "docker-build",
      "platform": "python",
      "dockerBuild": {
        "tag": "rapid:latest",
        "dockerfile": "${workspaceFolder}/Dockerfile",
        "context": "${workspaceFolder}",
        "pull": true
      }
    },
    {
      "type": "docker-run",
      "label": "docker-run: debug",
      "dependsOn": ["docker-build"],
      "dockerRun": {
        "containerName": "YOUR_IMAGE_NAME",
        "image": "YOUR_IMAGE_NAME:YOUR_IMAGE_TAG",
        "env": {
          "KEY1": "VAL1",
          "KEY2": "VAL2"
        },
        "volumes": [
          {
            "containerPath": "/app/data/out",
            "localPath": "${workspaceFolder}/data/out"
          },
          {
            "containerPath": "/mnt/in/input.csv",
            "localPath": "/tmp/file.csv"
          }
        ],
        "envFiles": ["../secret.env"],
        "ports": [
          {
            "containerPort": 5000,
            "hostPort": 5000
          }
        ]
      },
      "python": {
        "file": "main.py"
      }
    }
  ]
}
```

This leads to a `docker run --env-file "../secret.env"` beeing executed.

**Comment by fredjeck** [comment]
I would also love to have this feature implemented.
Currently handling multiple services which are relying on API keys, or certificates passwords which need to be passed to the docker build and debug tasks as env variables. I want to keep those out of Git and I want that each developer has its own local configuration. It would be much cleaner if this configuration can be defined in a local env file rather than having to put those infos into env scoped environment variables

**Comment by gustavorps** [comment]
+1

**Comment by Materie** [comment]
Still very much needed for the case where you want to not commit the env var data, but do want to commit the task itself.

I would like to have a task everyone in the team can reuse, then let them tweak the actual URIs where the task would run at by setting the env vars to locations on their machine which would be different for each developer.

**Comment by verikono** [comment]
5 Years Later.... (joke aside , was hoping this would be a feature by now) - keep up the good work, I love VSCode ;)

**Comment by ant1fact** [comment]
Please implement this. Currently the alternative is to add a project-specific settings.json that can hold config variables. This is one additional place where variables need to be maintained and they also don't provide any value outside of other vscode specific configuration files.

**Comment by tvsbrent** [comment]
Another workaround to share, although it would only work in a Linux / Mac environment, which worked fine for me, as I was intending this for a devcontainer.

Anyway, I created the following bash script to set the environment variables in the current context and then run the remaining positional arguments as a command.

```bash
#!/usr/bin/env bash

if [[ $# -lt 2 ]]
then
  echo "Usage: $(basename "$0") env_file (additional argument ...)"
  exit 1
fi

# We expect the first argument to be the path to the env file, so set those
# variables in the current shell and slice that argument off.
# shellcheck source=/dev/null
source "$1"
shift

exec "$@"
```

That allowed me to do this in the `tasks.json` file.
```json
{
  "tasks": [
    {
      "label": "test",
      "type": "shell",
      "command": "${workspaceFolder}/.devcontainer/run-task.sh",
      "args": [
        "env.tests",
        "/usr/local/python/current/bin/py.test",
        "/src",
        "--tb=line"
      ]
    }
  ]
}
```

**Comment by adriantorrie** [comment]
> Please implement this. Currently the alternative is to add a project-specific settings.json that can hold config variables. This is one additional place where variables need to be maintained and they also don't provide any value outside of other vscode specific configuration files.

Agree. Every other function in vscode supports a `.env` file directly.

Please implement support for `.env` in `tasks.json`

**Comment by adamrybak** [comment]
Would love this feature!

**Comment by eriktelepovsky** [comment]
+1

**Comment by PhilippPolterauer** [comment]
+1

**Comment by Unintendedz** [comment]
This feature would be extremely useful, The current workaround solutions are not ideal, either hardcoding sensitive information in the configuration files or using additional scripts to load the environment variables.

Being able to directly reference the .env file in tasks.json would allow us to better manage environment variables, while also keeping sensitive information out of the Git repository. This would greatly improve the UX of VSCode.

**Comment by emircanerkul** [comment]
+1

Wow, it was 7 years old issue. Actually similar was done for the launch one, didn't get why this is still on stale.

**Comment by doomleika** [comment]
I suppose this is not going to be done. ever since I've commented and subb'd this issue. Probably going to find an extension for this...

**Comment by LePau** [comment]
> > For the time being I am working around this by adding a prefix to my task command which essentially sources the `.env` file used by `launch.json`.
> 
> how do you do it? my .env file doesn't include the "export" keyword and thus `. .env` doesn't work as child processes can't see the variables. I also don't see any parameters to the `source` command to automatically export

Not OP, but this is my workaround for now.

> npm i -g dotenv-cli

Then in `tasks.json`:

```
        {
            "label": "build",
            "command": "dotenv",
            "type": "process",
            "args": [
                "-e",
                "${workspaceFolder}/.env",
                "--",
                "dotnet",
                "build",
                "${workspaceFolder}/mySolution.sln",
                "/property:GenerateFullPaths=true",
                "/consoleloggerparameters:NoSummary;ForceNoAlign"
            ],
```

**Comment by mathisgauthey** [comment]
I would really benefit from it on `launch.json` and `tasks.json` for the reasons stated [here](https://github.com/microsoft/vscode/issues/36769#issuecomment-495437733).

**Comment by lyf2000** [comment]
dn why downvoted for @joundso 's comment, but  
```
			"dockerRun": {
				"envFiles": [
					"${workspaceFolder}/borya.env"
				]
			}
```
worked for me, ty

full file .vscode/tasks.json:
```
{
	"version": "2.0.0",
	"tasks": [
		{
			"type": "docker-build",
			"label": "docker-build",
			"platform": "python",
			"dockerBuild": {
				"tag": "sonya:latest",
				"dockerfile": "${workspaceFolder}/Dockerfile",
				"context": "${workspaceFolder}",
				"pull": true
			}
		},
		{
			"type": "docker-run",
			"label": "docker-run: debug",
			"dependsOn": [
				// "docker-build"
			],
			"python": {
				"file": "main.py",
			},
			"dockerRun": {
				"envFiles": [
					"${workspaceFolder}/borya.env"
				]
			}
		}
	]
}
```

**Comment by sagarvelankar** [comment]
My solution is to change command `yarn dev` to `export $(grep -v -e '^\\s*#' .env | xargs) && yarn dev`.
Same should work for `npm`.

```.vscode/tasks.json
{
  "tasks": [
    {
      "command": "export $(grep -v -e '^\\s*#' .env | xargs) && yarn dev",
      "group": {
        "isDefault": true,
        "kind": "build"
      },
      "label": "yarn dev",
      "type": "shell"
    },
    {
      "command": "yarn install",
      "group": {
        "kind": "build"
      },
      "label": "yarn install",
      "type": "shell"
    },
  ],
  "version": "2.0.0"
}
```

**Comment by laverdet** [comment]
@sagarvelankar just source the variables:

```
{
	"version": "2.0.0",
	"options": {
		"env": {
			"ENV_FILE": "${workspaceFolder}/.env",
		},
	},
	"tasks": [
		{
			"label": "yarn dev",
			"type": "shell",
			"command": "set -a; . \"$ENV_FILE\"; yarn dev",
		},
	],
}
```

**Comment by igiona** [comment]
> ; . \"$ENV_FILE\"

This approach its nice, but would be even better if VsCode would allow to add a "env file" either in `options` or in the `task` itself.

**Comment by baebranch** [comment]
+1. This would be so awesome.

**Comment by lex-ibm** [comment]
Pretty please?

**Comment by lex-ibm** [comment]
I mean, it's been 7 years, 4 months and 17 days and people are still asking for this...

**Comment by starball5** [comment]
@rhezaas Not everyone uses .bashrc. Some people use .profile or the other bash startup file. You could handle this with some chained conditional logic.

**Comment by erhant** [comment]
why is this still not implemented after this many years, its such a natural use-case!

**Comment by tonkolviktor** [comment]
Hi, are there any updates here, looks like @zean11 and @starball5 are against this feature, but it's unclear why or is it a priority issue? In that case maybe someone could give out some pointers on how to implement this, so devs from the community could look into it...

**Comment by ferenci84** [comment]
It's interesting that several AI models recognize options.envFile as a valid option in task definitions.

I sent the following prompt to several AI models:

---
```json
{
	"label": "SSH DB server 123.123.123.123",
	"type": "shell",
	"command": "ssh 123.123.123.123",
},
```
Alter this task in vscode so that we have an envfile .env.ssh with a var PATH_TO_KEYFILE that should be given with the -i option of the ssh command.

---

The following models gave me an answer using the invalid options.envFile:
**Claude 4 sonnet**
**Qwen 3**
**o3 mini**

They all assumed that a configuration similar to the following one is valid:
```json
{
      "label": "SSH DB server 123.123.123.123",
      "type": "shell",
      "options": {
        "envFile": "${workspaceFolder}/.env.ssh"
      },
      "command": "ssh",
      "args": [
        "-i", "${env:PATH_TO_KEYFILE}",
        "123.123.123.123"
      ],
    }
```

**o3 mini** also wrote, "Here’s one way to do it, using the new 2.0.0 schema and the (undocumented) `options.envFile` support in tasks.json."

**Claude Opus 4** explicitly stated in its thinking that `envFile` is not natively supported by vs code and gave the solution with sourcing the env file, something like this:

```
{
			"label": "SSH DB server 123.123.123.123",
			"type": "shell",
			"command": "bash",
			"args": [
				"-c",
				"source .env.ssh && ssh -i \"$PATH_TO_KEYFILE\" root@123.123.123.123"
			],
		},
```

**DeepSeek R1** gave the same solution after several minutes and thinking of 5K tokens. In its thinking, I see that it's aware that the envFile property is not valid; however, in another shot, it gave me the wrong answer.

**Google Gemini 2.5** suggested using the extension https://marketplace.visualstudio.com/items?itemName=mikestead.dotenv

I just wrote this as an interesting fact.

Conclusion (for maintainers and contributors): 1. It's really a good idea to add this feature. 2. If you add this feature, it's a good idea to utilize the solution above, which was (incorrectly) agreed upon by numerous AI assistants as the best answer to this problem.

Workaround (for those looking for this feature): Until it is added (hopefully soon), use the above solution with `source` or the dotenv extension.

**Comment by tschifftner** [comment]
> [@sagarvelankar](https://github.com/sagarvelankar) just source the variables:
> 
> ```
> {
> 	"version": "2.0.0",
> 	"options": {
> 		"env": {
> 			"ENV_FILE": "${workspaceFolder}/.env",
> 		},
> 	},
> 	"tasks": [
> 		{
> 			"label": "yarn dev",
> 			"type": "shell",
> 			"command": "set -a; . \"$ENV_FILE\"; yarn dev",
> 		},
> 	],
> }
> ```

Works great, but I like this style a litte more:

```json
{
  "version": "2.0.0",
  "tasks": [
    {
      "label": "yarn dev",
      "type": "shell",
      "command": "set -a; source ${workspaceFolder}/.env; yarn dev"
    }
  ]
}
```
